Strange, this test passed on main and I have no idea why it would fail now:

```
/home/buildnode/jenkins/workspace/swift-PR-Linux@2/branch-release/5.5/swift/test/decl/protocol/req/recursion.swift:52:8: error: expected error not produced
17:43:41     // expected-error@-1{{conflicting arguments to generic parameter 'T' ('A.T' (associated type of protocol 'P') vs. 'S<A>')}}
17:43:41 ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
17:43:41 /home/buildnode/jenkins/workspace/swift-PR-Linux@2/branch-release/5.5/swift/test/decl/protocol/req/recursion.swift:58:28: error: incorrect message found
17:43:41     // expected-error@-1 {{conflicting arguments to generic parameter 'T' ('S<A>' vs. 'A.T' (associated type of protocol 'P')}}
17:43:41                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
17:43:41                            cannot convert value of type 'S<A>' to expected argument type 'A.T'
17:43:41 
17:43:41 --
```

--- a/lib/Sema/TypeCheckDeclPrimary.cpp
+++ b/lib/Sema/TypeCheckDeclPrimary.cpp
@@ -1563,7 +1563,7 @@ public:
       // Force some requests, which can produce diagnostics.
 
       // Check redeclaration.
-      (void) evaluateOrDefault(decl->getASTContext().evaluator,
+      (void) evaluateOrDefault(Context.evaluator,
                                CheckRedeclarationRequest{VD}, {});
 
       // Compute access level.
@@ -1577,8 +1577,11 @@ public:
       (void) VD->isObjC();
       (void) VD->isDynamic();
 
-      // Check for actor isolation.
-      (void)getActorIsolation(VD);
+      // Check for actor isolation of top-level and local declarations.
+      // Declarations inside types are handled in checkConformancesInContext()
+      // to avoid cycles involving associated type inference.
+      if (!VD->getDeclContext()->isTypeContext())
+        (void) getActorIsolation(VD);
 
       // If this is a member of a nominal type, don't allow it to have a name of
       // "Type" or "Protocol" since we reserve the X.Type and X.Protocol
@@ -1589,7 +1592,7 @@ public:
            VD->getName().isSimpleName(Context.Id_Protocol)) &&
           VD->getNameLoc().isValid() &&
           Context.SourceMgr.extractText({VD->getNameLoc(), 1}) != "`") {
-        auto &DE = getASTContext().Diags;
+        auto &DE = Context.Diags;
         DE.diagnose(VD->getNameLoc(), diag::reserved_member_name,
                     VD->getName(), VD->getBaseIdentifier().str());
         DE.diagnose(VD->getNameLoc(), diag::backticks_to_escape)

--- a/lib/Sema/TypeCheckProtocol.cpp
+++ b/lib/Sema/TypeCheckProtocol.cpp
@@ -5846,6 +5846,13 @@ void TypeChecker::checkConformancesInContext(IterableDeclContext *idc) {
   // Check all conformances.
   groupChecker.checkAllConformances();
 
+  // Check actor isolation.
+  for (auto *member : idc->getMembers()) {
+    if (auto *valueDecl = dyn_cast<ValueDecl>(member)) {
+      (void)getActorIsolation(valueDecl);
+    }
+  }
+
   if (Context.TypeCheckerOpts.DebugGenericSignatures &&
       !conformances.empty()) {
     // Now that they're filled out, print out information about the conformances

--- a/test/Concurrency/actor_isolation_cycle.swift
+++ b/test/Concurrency/actor_isolation_cycle.swift
@@ -1,0 +1,11 @@
+// RUN: %target-typecheck-verify-swift
+
+public protocol P {
+  associatedtype T
+  @MainActor func f(_: T)
+  @MainActor func g(_: T)
+}
+public struct S : P {
+  public func g(_: Int) {}
+  public func f(_: T) {}
+}

--- a/test/ModuleInterface/actor_isolation.swift
+++ b/test/ModuleInterface/actor_isolation.swift
@@ -68,6 +68,18 @@ public class C7 { }
 @available(macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0, *)
 extension C7: UnsafeSendable { }
 
+@available(macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0, *)
+public protocol P2 {
+  @SomeGlobalActor func method()
+}
+
+@available(macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0, *)
+// CHECK: class {{(Test.)?}}C8 : {{(Test.)?}}P2 {
+public class C8 : P2 {
+  // CHECK: @{{(Test.)?}}SomeGlobalActor public func method()
+  public func method() {}
+}
+
 // FIXME: Work around a bug where module printing depends on the "synthesized"
 // bit in conformances which is not serialized and not present in the textual
 // form.

--- a/test/decl/protocol/req/recursion.swift
+++ b/test/decl/protocol/req/recursion.swift
@@ -49,13 +49,12 @@ public struct S<A: P> where A.T == S<A> { // expected-error {{circular reference
 // expected-note@-3 {{while resolving type 'S<A>'}}
   func f(a: A.T) {
     g(a: id(t: a)) // `a` has error type which is diagnosed as circular reference
-    // expected-error@-1{{conflicting arguments to generic parameter 'T' ('A.T' (associated type of protocol 'P') vs. 'S<A>')}}
     _ = A.T.self
   }
 
   func g(a: S<A>) {
     f(a: id(t: a))
-    // expected-error@-1 {{conflicting arguments to generic parameter 'T' ('S<A>' vs. 'A.T' (associated type of protocol 'P')}}
+    // expected-error@-1 {{cannot convert value of type 'S<A>' to expected argument type 'A.T'}}
     _ = S<A>.self
   }
 

